### ISSUE TITLE
dpkg/dpkg_src re-design / update and security

### ISSUE THREAD

**Issue by gebi** [GoogleContainerTools/distroless#72]
Hi,

after a short look into dpkg/dpkg_src i'm still not sure how they are thought to be used (securely).

They do not validate signatures, they do not check the hash of packages and download from http, which all in all excludes them IMHO from use on production infrastructure.

Additionally i find it strange, that a core part of distroless now creates non-repeatable builds as neither a fixed version number is used nor a hash checked.

Thus i propose to split out what dpkg/dpkg_src currently does in skylark rules into an external script which just modifies/updates the WORKSPACE file to update packages and uses normal http_file declarations.
This would fix all the problems introduced with the current dpkg/dpkg_src rules and further clearly show what changed with debian dependencies in the repository diff.

This external tool then could also properly validate the signatures and provide secure download of debian packages with a sha256 sum it should automatically include (i can provide help with this if you want)

An additional nice feature would be to add an additional link to the same debian package in the http_file rule to point to http://snapshot.debian.org/, because mirrors typically don't hold on old packages for long. 

btw... if it's really the case that currently no security check/validation is done on the included debian packages what so ever, then the last versions of distroless images currently on gcr.io should be considered compromised and be removed ASAP.

**Comment by r2d4** [comment]
I think this could be simply fixed by just changing the source to point to a snapshot and checking the hash.  Additionally, each .deb can be checked against the hash included in the Packages.gz file.

**Comment by mattmoor** [comment]
@r2d4 we absolutely need to check the package against what's in the Packages.gz.

**Comment by r2d4** [comment]
@gebi I added checksum validation to the `dpkg_src` rule as well as checksum validation while downloading the .debs from the Packages.gz provided checksum.  I also switched it over to use the snapshot archives.  I think that should resolve this issue?

https://github.com/GoogleCloudPlatform/distroless/pull/71

**Comment by gebi** [comment]
It fixes the special case for debian because there is a snapshot.debian.org, all other package sources still have the problem.

It also removes the possibility to fix problems in debian as far as i can see because it just uses packages from the snapshot on snapshot.debian.org.

All in all it seems we will keep our fork for the forseeable future and explore the possibility to upgrade the WORKSPACE file with helper tools instead of using the new the dpkg/dpkg_src rules.

**Comment by mattmoor** [comment]
@gebi Generally updating `WORKSPACE` directly is hard because we don't want to trample on user edits; however, @r2d4 I think we could reasonably generate a `.bzl` file based on the package mirror and some specification (with the validation @gebi) suggests and check it in.  

Then in `WORKSPACE` (among other user code) we'd include the generated imports via:
```python
load("//path/to:generated.bzl", "dpkg_imports")
dpkg_imports("name-to-give-the-repository")
```

Ultimately, (and @gebi correct me if I'm wrong) the goal is to be able to float dependencies tracking some mutable label, but check in the Bazel equivalent to a "lock file" (e.g. `package-lock.json`, `yarn.lock`, `requirements.txt.lock`)

**Comment by mattmoor** [comment]
Frankly, I wonder if Bazel needs a `WORKSPACE.lock` concept...

**Comment by MarkusTeufelberger** [comment]
Tools like gazel or gazelle also automatically generate BUILD files and also update them, if necessary. You could set tags to rules to prevent them from being updated or to indicate that your tool is processing them automatically.

If you also have an external "lock" file, you risk that the version defined in the external file does not match the version in the WORKSPACE. IMHO a better approach would be a tool that parses WORKSPACE files for rules with a certain tag and suggests possible updates and allows adding packages there, to keep the WORKSPACE as a single source of truth for external package files.

Your current solution tries to keep everything inside Bazel while using debian mirrors as external source of truth (aka "which version gets installed"). I would prefer to keep the source of truth inside the WORKSPACE and have a non-Bazel tool to interact with it. This seems to be more in line with using Bazel as build tool, not as dependency management or resolution platform.

**Comment by mattmoor** [comment]
@MarkusTeufelberger That is a fair point, but I doubt everyone looking to build a repository rule is looking to build something like gazel either.  The same argument about drifting is true of Gazel[le] too if I don't run it (not missing deps because the build will break, but extra deps).

I think the fact that all we have today is `WORKSPACE` means folks are left using it to both resolve as well as fetch their dependencies.  I agree that the `WORKSPACE` input to Bazel should be fully resolved and just a fetch of checksum'd or content-addressed things, but that's a maintenance nightmare so we need a way for folks to express what floating HEAD of which that's a snapshot.

The question is: what is the input to the external tool that generates `WORKSPACE`?  I don't know that we have a strong implicit signal we can key off of like Gazel[le] + Go sources, so we want a high-level description similar to what's here, but instantiated to something with the strength of checksums.  

This problem of label => hash applies to every repository rule, so rather than needing N Gazel[le]'s, solving this at the Bazel level seems appropriate.


This is why I'm tempted to make the extreme statement: `WORKSPACE` => `WORKSPACE.lock`** resolution should become a separate tool from Bazel (like Gazel[le]), and Bazel should key off of `WORKSPACE.lock`.  Perhaps Bazel can indefinitely fallback on running the tool when `WORKSPACE.lock` doesn't exist.

** - If you are married to `WORKSPACE` as the Bazel input then `DEPENDENCIES` => `WORKSPACE`

**Comment by MarkusTeufelberger** [comment]
> what is the input to the external tool that generates WORKSPACE?

User decisions. E.g. "I want to use libfoo 0.1.2 and libbar 1.3.3.7 - they should be available at one of the following 3 Debian mirrors: [1, 2, 3]". This is enough input to generate reproducible WORKSPACE files with checksums and everything. On subsequent runs you can also parse WORKSPACE files (e.g. with buildozer) and suggest updates for the user to confirm.

Not as sexy as a Bazel internal dpkg() rule with some magic, but easier to extend for other distros and it is always very clear at every point in time what software is going to end up in your containers (at the moment you seem to download whatever a mirror serves you over plain http without even checking the signature of the checksum file, which is also not used).

**Comment by mattmoor** [comment]
> User decisions.

Right, we're saying essentially the same thing.

You seem to be suggesting that both the user intent as well as its point-in-time instantiation live in `WORKSPACE` and a tool exists to update the latter based on the former.  I'm simply stating that another way would be to have a "floating HEAD" version of `WORKSPACE` and a "lock" version of `WORKSPACE` with a tool to create the latter from the former.

**Comment by MarkusTeufelberger** [comment]
The "floating HEAD" version would not need to be formatted like a WORKSPACE file at all, since one would anyways always build from a "lock" version (a "floating HEAD" would be just locked to the most recent available version at the time the build is started).

I would love to be able to just specify for example a whole git repo as external dependency and then in the "lock" file specify the commit hash to be used. Currently however Bazel doesn't work that way. This project at the moment only uses "floating HEAD" which is a quite uncommon configuration. Especially when done in such an insecure way. Pinning the global state at one point in time (the proposed change to dpkg_src) is a nice first step to get back to what Bazel currently is good at, however pinning individual versions seems preferable to me.

Anyways: If you have a "floating" and a "locked" version (which can be updated from the "floating" version), the "floating" file doesn't need to have anything to do with Bazel at all. After all, the tagline of Bazel is "{Fast, Correct} - Choose two" - and _only_ reproducible builds are correct. I don't see any reason why Bazel should deal with "floating HEAD" WORKSPACE files in the first place, other than to potentially warm up caches for an eventually occurring correct build in the future.

Since the "lock" version also needs to contain enough information to generate a "floating HEAD" version from it (package names and mirror URLs), but also contains additional information (the exact version/hash to use), I would argue that it is more beneficial to focus on the "lock" version of rules than the more relaxed "floating HEAD" ones.

Tl;dr: It is easy to also get a current version if you already know an exactly defined version, it is not possible to define an exact version if you only can define current versions (or frozen sets of versions). This is why it is more important to be able to define exact versions as granular as possible.

**Comment by dlorenc** [comment]
I'm trying to catch up on this discussion a bit, so apologies if I'm misunderstanding anything in here.

I think we have one main fundamental constraint here - the contents of a set of WORKSPACE/BUILD files should result in the same build each time, or fail if the external artifacts are no longer available/have changed/fail.

The current design works by having users pin a specific Packages.gz file (by URL, which introduces the implicit requirement that this never changes), and then installs whatever .deb versions happen to be in that Packages.gz.

I can imagine another API that lets users specify a set of debian package list (by URL, by file, whatever).  These would not need to be pinned by SHA.

Then separately, users would specify a list of packages to install, these would be specified by SHA/version.

The bazel rules would then need to search the specified package lists for the package versions that the user has requested, and either download them or fail if they are no longer available.

I *think* this will lead to general frustration when trying to use the standard debian mirrors, since they make no promises that specific package versions will remain available. So to make this usable you would probably still need to use a local package mirror that does keep old versions around or use the snapshot mirrors.

But perhaps this is what #96 is trying to accomplish?

**Comment by donmccasland** [comment]
Closing old bug.  Please reopen if still valid.

**Comment by MarkusTeufelberger** [comment]
Do you verify PGP signatures now and hash the downloaded files?
